delphi/latents/neighbours.py

```diff
@@ -20,148 +21,163 @@ class NeighbourCalculator:
         # Compute the similarity between latents
-        similarity_matrix = weight_matrix_normalized.T @ weight_matrix_normalized
-
-        # Get the indices of the top k neighbours for each latent
-        top_k_indices = torch.topk(
-            similarity_matrix, self.number_of_neighbours, dim=1
-        ).indices
-
-        # Return the neighbour lists
-        return {i: top_k_indices[i].tolist() for i in range(len(top_k_indices))}
-
-    def _compute_correlation_neighbours(self) -> dict[int, list[int]]:
+        done = False
+        batch_size = weight_matrix_normalized.shape[0]
+        number_latents = batch_size
+
+        neighbour_lists = {}
+        while not done:
+            try:
+                for start in tqdm(range(0, number_latents, batch_size)):
+                    rows = wT[start : start + batch_size]
+                    similarity_matrix = weight_matrix_normalized @ rows
+                    indices, values = torch.topk(
+                        similarity_matrix, self.number_of_neighbours + 1, dim=1
+                    )
+                    neighbour_lists.update(
+                        {
+                            i
+                            + start: list(
+                                zip(indices[i].tolist()[1:], values[i].tolist()[1:])
+                            )
+                            for i in range(len(indices))
+                        }
+                    )
+                    del similarity_matrix
+                    torch.cuda.empty_cache()
+                done = True
+            except Exception:
+                batch_size = batch_size // 2
+                if batch_size < 2:
+                    raise ValueError(
+                        "Batch size is too small to compute similarity matrix. "
+                        "You don't have enough memory."
+                    )
+
+        return neighbour_lists
+
+    def _compute_cooccurrence_neighbours(self) -> dict[int, list[tuple[int, float]]]:
         """
-        Compute neighbour lists based on activation correlation patterns.
+        Compute neighbour lists based on feature co-occurrence in the dataset.
+        If you run out of memory try reducing the token_batch_size
+        Code adapted from https://github.com/taha-yassine/SAE-features/blob/main/cooccurrences/compute.py
         """
 
-        # the preactivation_matrix has the shape (number_of_samples,hidden_dimension)
-        preactivation_matrix = self.pre_activation_record
-
-        # compute the covariance matrix of the preactivation matrix
-        covariance_matrix = torch.cov(preactivation_matrix.T)
-
-        # load the encoder
-        encoder_matrix = self.autoencoder.encoder.weight
-
-        # covariance between the latents is u^T * covariance_matrix * u
-        covariance_between_latents = (
-            encoder_matrix.T @ covariance_matrix @ encoder_matrix
-        )
+        import cupy as cp
+        import cupyx.scipy.sparse as cusparse
```

Comment:
do we really need to rely on CuPy for this? I've found that it can sometimes fight with PyTorch for control over the CUDA version

Torch has limited support for CSC tensors on CUDA